### PR TITLE
Add windows auto config url

### PR DESCRIPTION
**Pull request by partis**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -9,7 +9,7 @@ before_install:
   - bundle --version
 bundler_args: --without=development
 
-rvm: 2.0.0
+rvm: 2.2.2
```
Comment (tmatilai) [inline .travis.yml]: Older Vagrant versions bundle Ruby 2.0 so we really want to test with that version as long as we want to support those Vagrant versions. And I don't think we want to cut the support unless needed.

If the build or tests are broken with the old ruby, we should fix that first. I can take a look later.

File: lib/vagrant-proxyconf/action/configure_env_proxy.rb [hunk 2]
```diff
@@ -45,10 +47,10 @@ def set_windows_system_proxy(proxy)
...
             path    = "HKCU\\Software\\Microsoft\\Windows\\CurrentVersion\\Internet Settings"
 
-            proxy1  = "cmd.exe /C reg add \"#{path}\" /v ProxyEnable   /t REG_DWORD /d 1                          /f"
-            proxy2  = "cmd.exe /C reg add \"#{path}\" /v ProxyServer   /t REG_SZ    /d #{config.http.inspect}     /f"
-            proxy3  = "cmd.exe /C reg add \"#{path}\" /v ProxyOverride /t REG_SZ    /d #{config.no_proxy.inspect} /f"
-            proxy4  = "cmd.exe /C reg add \"#{path}\" /v AutoDetect    /t REG_DWORD /d 0                          /f"
+            proxy1  = "cmd.exe /C reg add \"#{path}\" /v ProxyEnable   /t REG_DWORD /d 1                            /f"
+            proxy2  = "cmd.exe /C reg add \"#{path}\" /v ProxyServer   /t REG_SZ    /d #{config.http.inspect}       /f"
+            proxy3  = "cmd.exe /C reg add \"#{path}\" /v ProxyOverride /t REG_SZ    /d #{config.no_proxy.inspect}   /f"
+            proxy4  = "cmd.exe /C reg add \"#{path}\" /v AutoDetect    /t REG_DWORD /d 0                            /f"
```
Comment (tmatilai) [inline lib/vagrant-proxyconf/action/configure_env_proxy.rb]: Are these changes needed? =)

File: lib/vagrant-proxyconf/action/configure_env_proxy.rb [hunk 3]
```diff
@@ -61,6 +63,40 @@ def set_windows_system_proxy(proxy)
           end
         end
 
+        def set_windows_auto_config(autoconfig)
+          if autoconfig
+            path    = "HKCU\\Software\\Microsoft\\Windows\\CurrentVersion\\Internet Settings"
+            proxy1  = "cmd.exe /C reg add \"#{path}\" /v AutoConfigURL /t REG_SZ    /d #{config.autoconfig.inspect} /f"
```
Comment (tmatilai) [inline lib/vagrant-proxyconf/action/configure_env_proxy.rb]: Maybe a better name than`proxy1`? `command` or alike?

File: lib/vagrant-proxyconf/action/configure_env_proxy.rb [hunk 4]
```diff
@@ -61,6 +63,40 @@ def set_windows_system_proxy(proxy)
...
+
+            logger.info('Setting system auto config settings')
+
+            @machine.communicate.sudo(proxy1)
+
+            set_windows_ie_settings
+
```
Comment (tmatilai) [inline lib/vagrant-proxyconf/action/configure_env_proxy.rb]: Extra newline

File: lib/vagrant-proxyconf/action/configure_env_proxy.rb [hunk 5]
```diff
@@ -61,6 +63,40 @@ def set_windows_system_proxy(proxy)
...
+            logger.info("Not setting auto config settings")
+          end
+        end
+
+        def set_windows_ie_settings
+          path    = "HKCU\\Software\\Microsoft\\Windows\\CurrentVersion\\Internet Settings\\Connections"
+          keys = ["DefaultConnectionSettings", "SavedLegacySettings"]
+
+          keys.each {
+            |key| connectionHex1 = "cmd.exe /C reg query \"#{path}\" /v #{key} /t REG_BINARY"
```
Comment (tmatilai) [inline lib/vagrant-proxyconf/action/configure_env_proxy.rb]:
```ruby
keys.each do |key|
  command = ...
  # ...
end
```

File: lib/vagrant-proxyconf/action/configure_env_proxy.rb [hunk 6]
```diff
@@ -61,6 +63,40 @@ def set_windows_system_proxy(proxy)
...
+            @machine.communicate.sudo(connectionHex1) do |type, data|
+              if type == :stdout
+                if data.include? key
+                  hex = update_hex(data.split()[2], 8, "05")
```
Comment (tmatilai) [inline lib/vagrant-proxyconf/action/configure_env_proxy.rb]: This is a bit to magic with those parameters. Maybe just pass the data to the method, and possibly rename the method to be more descriptive.

File: lib/vagrant-proxyconf/action/configure_env_proxy.rb [hunk 7]
```diff
@@ -117,6 +153,12 @@ def environment_config
             no_proxy=#{config.no_proxy || ''}
           CONFIG
         end
+
+        def update_hex(hex, index, value)
```
Comment (tmatilai) [inline lib/vagrant-proxyconf/action/configure_env_proxy.rb]: This method also needs some comments explaining what it does, unless it is clear from the method name.

Bonus points for unit tests.

File: lib/vagrant-proxyconf/config/env_proxy.rb [hunk 8]
```diff
@@ -46,12 +49,22 @@ def config_for(key, value)
...
+            # Quoting `autoconfig` value as its a url
+            value = value.inspect if key.name == :autoconfig
+
             [var.upcase, var.downcase].map { |v| "export #{v}=#{value}\n" }.join
           end
         end
 
         def env_variable_name(key)
-          key.name == :no_proxy ? "no_proxy" : "#{key.name}_proxy"
+          #key.name == :no_proxy ? "no_proxy" : "#{key.name}_proxy"
```
Comment (tmatilai) [inline lib/vagrant-proxyconf/config/env_proxy.rb]: Please just remove the line. It can be found from git history if needed.

File: lib/vagrant-proxyconf/config/env_proxy.rb [hunk 9]
```diff
@@ -46,12 +49,22 @@ def config_for(key, value)
...
+          if key.name == :no_proxy
+            return "no_proxy"
```
Comment (tmatilai) [inline lib/vagrant-proxyconf/config/env_proxy.rb]: As the `if` is the last statement of the method, the `return` can be dropped.

I would also use a `case` statatement here:

```ruby
case key.name
when :no_proxy
  "no_proxy"
when :autoconfig
  "auto_config_url"
else
  "#{key.name}_proxy"
end
```

File: .travis.yml (hunk 1 above)
Comment (partis) [inline .travis.yml]: When the test were ran using 2.0.0 vagrant version 1.8.1 passed but the others failed. Looking through it versions 1.7.2 and 1.6.5 failed because the ruby version wasn't high enough for the version of rack being installed (2.0.3) and for the older versions it wasn't high enough for the version of nokogiri being installed (1.7.2) and i'd assume rack as well if it got to that point.

I'd figured it was pulling the versions selected from the vagrant version so not something that could be controlled from here.

Comment (tmatilai) [inline .travis.yml]: I hope we can pin down some transitive dependencies. I'll fix the build when I find the time.

Thanks anyway for the PR!

File: lib/vagrant-proxyconf/action/configure_env_proxy.rb (hunk 2 above)
Comment (partis) [inline lib/vagrant-proxyconf/action/configure_env_proxy.rb]: no, the spaces are from an earlier iteration where i set the config url alongside these, will remove them

File: lib/vagrant-proxyconf/action/configure_env_proxy.rb (hunk 3 above)
Comment (partis) [inline lib/vagrant-proxyconf/action/configure_env_proxy.rb]: Thanks will update.

File: lib/vagrant-proxyconf/action/configure_env_proxy.rb (hunk 5 above)
Comment (partis) [inline lib/vagrant-proxyconf/action/configure_env_proxy.rb]: Thanks fairly new to Ruby so always happy to get better ways of doing things

File: lib/vagrant-proxyconf/action/configure_env_proxy.rb (hunk 6 above)
Comment (partis) [inline lib/vagrant-proxyconf/action/configure_env_proxy.rb]: Ok, thanks

**Comment by tmatilai** [comment]
@partis I finally got to fix the build matrix. So if you could remove all changes from there and (squash and) rebase your work on top of current master, I can take another look at this.

**Comment by codylane** [comment]
Old and has conflicts that were never resolved.  Closing issue.
